File: leetcode/cpp/1140.stone-game-ii.cpp
class Solution {
public:
    int stoneGameII(vector<int>& piles) {
        int n = piles.size();
        vector<vector<int>> memo(n, vector<int>(n));
        vector<int> suffix = piles;
        for (int i=n-2; i > -1; i--) {
            suffix[i] += suffix[i+1];
        }
        return stoneGameII(memo, suffix, 0, 1);
    }
private:
    int stoneGameII(vector<vector<int>>& memo, vector<int>& suffix, int i, int m) {
        if (i + 2 * m >= suffix.size()) {
            return suffix[i];
        }
        if (memo[i][m] == 0) {
            int bob = suffix[i];
            for (int X=1; X <= 2 * m; X++) {
                bob = min(bob, stoneGameII(memo, suffix, i + X, max(m, X)));
            }
            memo[i][m] = suffix[i] - bob;
        }
        return memo[i][m];
    }
};

File: leetcode/go/1140.stone-game-ii.go
func stoneGameII(piles []int) int {
    n := len(piles)
    memo := make([][]int, n)
    for i := range n {
        memo[i] = make([]int, n)
    }
    suffix := make([]int, n)
    copy(suffix, piles)
    for i:=n-2; i > -1; i-- {
        suffix[i] += suffix[i+1]
    }
    return dp(&memo, &suffix, 0, 1)
}

func dp(memo *[][]int, suffix *[]int, i int, m int) int {
    if i + 2 * m >= len((*suffix)) {
        return (*suffix)[i]
    }
    if (*memo)[i][m] == 0 {
        bob := (*suffix)[i]
        for X:=1; X < 2*m + 1; X++ {
            bob = min(bob, dp(memo, suffix, i+X, max(X, m)))
        }
        (*memo)[i][m] = (*suffix)[i] - bob
    }
    return (*memo)[i][m]
}

